platform/sensor-stream-manager/sensor-stream-manager.js

```diff
@@ -1,72 +1,208 @@
+    }
+}
+
+/**
+ * This function sends sensor requirement to the remote gateway.
+ * @param {string} ip -remote gateway's ip
+ * @param {string[]} sensorIds - an array of sensor id
+ * @param {string} topic - application's topic
+ */
+function registerToRemoteGateway(ip, sensorIds, topic) {
+    // Remote gateway's register-topic url
+    const gatewayUrl = `http://${ip}:5000/gateway/register-app-sensor-reqruirement`;
+    // Request body
+    const body = {
+        ip: gatewayIp,
+        sensors: sensorIds,
+        topic: topic,
+    };
+    // Send application's sensor requirement to remote gateway
+    fetch(gatewayUrl, {
+        method: 'POST',
+        body: JSON.stringify(body),
+        headers: {'Content-Type': 'application/json'},
+        timeout: 5000
+    }).then(res => {
+        if(res.status === 200) {
+            console.log(`[INFO] Sent "${topic}" to ${ip} successfully!`);
+        } else {
+            console.error(`[ERROR] Failed to send "${topic}" to ${ip} with status ${res.status}.`);
+        }
+    }).catch(err => {
+        console.error(`[ERROR] Failed to send "${topic}" to ${ip}.`);
+        console.error(err);
+    });
+}
+
+/**
+ * This function register MQTT clients.
+ * @param {string} ip - MQTT broker's ip
+ */
+function registerMQTTClient(ip) {
+    // Check if the MQTT client exists
+    if(!mqttClients[ip]) {
+        const client = connectToMQTTBroker(ip);
+        client.on('connect', () => {
+            if(ip === gatewayIp) {
+                addLocalGatewayDataEvent(client);
+            }
+            client.on('disconnect', () => {
+                console.log(`[INFO] Disconnected to MQTT broker at ${ip}.`);
+            });
+            console.log(`[INFO] Connected to MQTT broker at ${ip} successfully!`);
+        });
+        mqttClients[ip] = client;
+    }
+}
+
+/**
+ * This function publishes data to application's topic.
+ * @param {string} ip - MQTT broker's ip
+ * @returns {Object} - MQTT client
+ */
+function connectToMQTTBroker(ip) {
+    const url = `mqtt://${ip}`;
+    // Connect to mqtt broker
+    return mqtt.connect(url);
+}
+
+/**
+ * This function publishes data to application's topic.
+ * @param {string} ip - MQTT broker's ip
+ * @param {string} topic - application's topic
+ * @param {string} data - sensor data
+ */
+function publishData(ip, topic, data) {
+    // Check if the MQTT client exists
+    if(!mqttClients[ip]) {
+        console.error(`[ERROR] ${ip} has not been registered.`);
+        return;
+    }
+    const client = mqttClients[ip];
+    client.publish(topic, data, {}, err => {
+        if(err) {
+            console.error(`[ERROR] Failed to publish to ${ip}.`);
+            console.error(err);
+        }
+    });
+}
+
+/**
+ * This function lets the local MQTT client
+ * subscribes to "gateway-data" topic and
+ * publishes sensor stream data to application's topic
+ * @param {Object} client - MQTT client
+ */
+function addLocalGatewayDataEvent(client) {
```

Comment:
I think this function needs a rewrite. 
Intuitively, I'm not getting the meaning of what the function does by reading the name "addLocalGatewayDataEvent". 
This function does the following things and can be split into two:
1. Subscribe to the gateway-data MQTT topic => function subscribeToTopic(client, topic)
2. Route sensorstreams to registered applications => function routeSensorStreamsToApps(client, sensorStream)